helix-core/src/main/java/org/apache/helix/controller/rebalancer/waged/algorithm/ConstraintBasedAlgorithm.java

```diff
@@ -0,0 +1,108 @@
+import org.apache.helix.controller.rebalancer.waged.model.ClusterContext;
+import org.apache.helix.controller.rebalancer.waged.model.ClusterModel;
+import org.apache.helix.controller.rebalancer.waged.model.OptimalAssignment;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * The algorithm is based on a given set of constraints
+ * - HardConstraint: Approve or deny the assignment given its condition, any assignment cannot
+ * bypass any "hard constraint"
+ * - SoftConstraint: Evaluate the assignment by points/rewards/scores, a higher point means a better
+ * assignment
+ * <p>
+ * The goal is to accumulate the most points(rewards) from "soft constraints" while avoiding all
+ * "hard constraints"
+ */
+public class ConstraintBasedAlgorithm implements RebalanceAlgorithm {
+  private static final Logger LOG = LoggerFactory.getLogger(ConstraintBasedAlgorithm.class);
+  private final List<HardConstraint> _hardConstraints;
+  private final List<SoftConstraint> _softConstraints;
+
+  public ConstraintBasedAlgorithm(List<HardConstraint> hardConstraints,
+      List<SoftConstraint> softConstraints) {
+    _hardConstraints = hardConstraints;
+    _softConstraints = softConstraints;
+  }
+
+  @Override
+  public OptimalAssignment calculate(ClusterModel clusterModel) {
+    ClusterContext clusterContext = clusterModel.getContext();
+    OptimalAssignment optimalAssignment = new OptimalAssignment(clusterContext);
+    Map<String, Set<AssignableReplica>> replicasByResource = clusterModel.getAssignableReplicaMap();
+    List<AssignableNode> nodes = new ArrayList<>(clusterModel.getAssignableNodes().values());
+
+    for (String resource : replicasByResource.keySet()) {
+      for (AssignableReplica replica : replicasByResource.get(resource)) {
+        Optional<AssignableNode> maybeBestNode =
+            getNodeWithHighestPoints(replica, nodes, optimalAssignment);
+        // stop immediately if any replica cannot find best assignable node
+        if (optimalAssignment.hasAnyFailure()) {
+          LOG.error(
+              "Unable to find any available candidate node for partition {}; Fail reasons: {}",
+              replica.getPartitionName(), optimalAssignment.getFailures());
+          return optimalAssignment;
+        }
+        maybeBestNode.ifPresent(node -> optimalAssignment.addAssignment(replica, node));
+      }
+    }
+
+    return optimalAssignment;
+  }
+
+  private Optional<AssignableNode> getNodeWithHighestPoints(AssignableReplica replica,
+      List<AssignableNode> assignableNodes, OptimalAssignment currentAssignment) {
+    List<AssignableNode> candidateNodes = assignableNodes.stream()
+        .filter(candidateNode -> _hardConstraints.stream().allMatch(constraint -> {
+          boolean isValid = constraint.isAssignmentValid(candidateNode, replica, currentAssignment.getClusterContext());
+          if (!isValid) {
+            currentAssignment.trackAssignmentFailure(candidateNode, replica, constraint);
```

Comment:
Is this necessary? We basically record all the possible pairs, right?
So even with a completely valid assignment, you will have (nodeCount - 1) * replicaCount error records.

I don't think it is necessary.